_posts/2018-12-04-imposing-invariants-in-scala.md

```diff
@@ -0,0 +1,353 @@
+---
+layout: post
+author: cmteixeira
+title: Imposing invariants in Scala
+---
+
+
+
+Imagine a opera house allows reservations to a show. For each reservation two people can come: The buyer of the reservation himself, and his guest.  
+As in any show-house, there are many types of seats; some are standing, some normal and some on vip tribunes.
+
+
+
+As a restriction from the opera house, both buyer and guest must be on the same seating area. It is not permitted for one ticket to be standing while the other for the normal seated area for example.  
+I believe this is what people call an invariant. A requirement that must hold for the data to have meaning.  
+How to enforce then this constraint ?  
+Let us first define the data types relevant to the problem:  
+
+
+```scala
+sealed trait Seat
+
+case class Seated(row: Int, column: Int) extends Seat
+case class Tribune(tribuneId: String) extends Seat
+case object Standing extends Seat
+```
+
+In what follows it will be useful to discern the type relation above on the following diagram.
+
+<img class=" size-full wp-image-397 aligncenter" src="https://aerodatablog.files.wordpress.com/2018/12/typehiearchy1.png" alt="ActionBuilder" />  
+
+We also have:  
+
+```scala
+case class Ticket[SeatType <: Seat](
+  verificationCode: String,
+  price: Int,
+  seat: SeatType
+)
+```
+
+```scala
+case class Reservation[SeatType <: Seat](
+  userTicket: Ticket[SeatType],
+  guestTicket: Ticket[SeatType]
+)
+```
+
+
+The first thing to notice is that the above does not deliver what we want. The following compiles:  
+
+```scala
+val reservation = Reservation(
+  usertTicket = Ticket("id1", 123, VipTribuneSeat("east")),
+  guestTicket = Ticket("id2", 123, Standing)
+)
+```
+
+The compiler tries to infer the type <code>SeatType</code>for reservation. From the user ticket on line 2 it thinks it is <code>VipTribuneSeat</code>. The compiler is satisfied since this a sub-type of <code>Seat</code> which is an upper type bound on Reservation. This is contradicted on the 3rd line, as we encounter type <code>Standing</code> for the guest ticket. Bad news since type <code>SeatType</code> only line 2 and <code>SeatType</code> on line 3 refer, well, to the same type!  
```

Comment:
Bad news since type <code>SeatType</code> **only** line

This `only` is confusing, did you mean `on` ?

Reply:
Yes.
Done